torch/csrc/jit/passes/onnx/remove_inplace_ops_for_onnx.cpp

```diff
@@ -567,256 +370,390 @@ Value* findArgumentAsInputParam(
+    auto* in = n->input(i);
+    auto* alias = vt_.findAliasForValueAtNode(in, n);
+
+    if (alias != in) {
+      n->replaceInput(i, alias);
+      GRAPH_UPDATE(
+          "Replacing ",
+          in->debugName(),
+          " with ",
+          alias->debugName(),
+          " for ",
+          *n);
     }
   }
 }
 
-// clang-format off
-// The registerInplaceOpAsBlockOutputs function tracks inplace op
-// (like aten::copy_ or aten::append) outputs as sub-block output.
-// Also, match the number of If sub-block outputs
-// if a value is updated in one branch, but no updated on the other branch.
-// For example:
-// = prim::If(%30)
-//    block0():
-//      ...
-//      %35 : Tensor = aten::copy_(%state_copy.1, %33, %12)
-//      -> ()
-//    block1():
-//      ...
-//      %40 : Tensor = aten::copy_(%state.1, %38, %12)
-//      -> ()
-//
-// After the pass
-//%_output_state_copy.1 : Tensor, %_output_state.1 : Tensor = prim::If(%30)
-//    block0():
-//      %_output_state.2 : Tensor = aten::clone(%state.1, %59)
-//      ...
-//      %_output_state_copy.3 : Tensor = onnx::Placeholder[name="index_put_"](%state_copy.1)...
-//      ...
-//      -> (%_output_state_copy.3, %_output_state.2)
-//    block1():
-//      %50 : None = prim::Constant()
-//      %_output_state_copy.2 : Tensor = aten::clone(%state_copy.1, %50)
-//      ...
-//      %_output_state.3 : Tensor = onnx::Placeholder[name="index_put_"](%state.1)...
-//       ...
-//      -> (%_output_state_copy.2, %_output_state.3)
-std::unordered_map<std::string, Value*> registerInplaceOpAsBlockOutputs(
+// Find the correct alias representing Value v at Node n.
+Value* InplaceConverter::ValueTracker::findAliasForValueAtNode(
+    Value* v,
+    const Node* n) const {
+  GRAPH_UPDATE("Finding alias for value:", v->debugName(), " at node ", *n);
+  if (alias_to_value_.find(v) == alias_to_value_.end()) {
+    // This value was not affected by any inplace operator.
+    return v;
+  }
+
+  auto* root_v = alias_to_value_.find(v)->second;
+  TORCH_INTERNAL_ASSERT(
+      value_to_sorted_aliases_.find(root_v) != value_to_sorted_aliases_.end());
+  const auto& aliases = value_to_sorted_aliases_.find(root_v)->second;
+
+  // alias is accessible only if
+  // 1. alias owning block is ancestor of n.
+  // 2. alias owning node is before n.
+  // return the last alias that satisfies this condition.
+  Value* found_alias = nullptr;
+  for (auto* alias : aliases) {
+    auto* alias_n = alias->node();
+    if (alias_n->isBefore(n) &&
+        isAncestor(alias_n->owningBlock(), n->owningBlock())) {
+      found_alias = alias;
+    }
+  }
+
+  TORCH_INTERNAL_ASSERT(nullptr != found_alias, "More details: \n",
+    n->sourceRange().str(),
+    "Input ", v->debugName(), " of node ", *n,
+    " was modified by in-place operation, but we cannot find its updated value. ",
+    "Please report a bug to PyTorch, and/or try to avoid using in-place operators on this value.");
```

Comment:
what situations can cause this? Why does the user need to report a bug?

Reply:
unknown cases :) This is not expected to happen, but if it does, the hope is that user can get some meaningful message explaining what might have gone wrong.